### PR TITLE
linuxPackages: unbreak new kernels on 32-bit platforms 

### PR DESCRIPTION
**Pull request by misuzu**
###### Description of changes
Disable BTF on 5.15+ for 32-bit platforms.
Related: https://github.com/NixOS/nixpkgs/pull/177401

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [X] i686-linux
  - [ ] aarch64-linux
  - [X] armv7l-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by misuzu** [comment]
@ofborg build pkgsi686Linux.linux_5_10 pkgsi686Linux.linux_5_15 pkgsi686Linux.linux_5_17 pkgsi686Linux.linux_5_18

**Review by NeQuissimus — APPROVED** [review]

File: pkgs/os-specific/linux/kernel/common-config.nix [hunk 1]
```diff
@@ -41,7 +41,7 @@ let
         (whenBetween "5.2" "5.18" yes)
       ];
       DEBUG_INFO_DWARF_TOOLCHAIN_DEFAULT = whenAtLeast "5.18" yes;
-      DEBUG_INFO_BTF            = whenAtLeast "5.2" (option yes);
+      DEBUG_INFO_BTF            = whenAtLeast "5.2" (option (if stdenv.hostPlatform.is32bit && (versionAtLeast version "5.15") then no else yes));
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/kernel/common-config.nix:44]: We should add a comment here explaining why 32 bit has it disabled

Comment (misuzu) [reply pkgs/os-specific/linux/kernel/common-config.nix:45]: > We should add a comment here explaining why 32 bit has it disabled

Done!

**Review by Mindavi — APPROVED** [review]

**Comment by martinetd** [comment]
That sounds like a pahole bug, was it reported upstream?
I don't have the resources to reproduce this now but it probably needs some follow up rather than just disabling...

**Comment by misuzu** [comment]
> That sounds like a pahole bug, was it reported upstream?
> I don't have the resources to reproduce this now but it probably needs some follow up rather than just disabling...

I'm not a kernel specialist, but that's probably involves some kind of mailing list? I'm simply want my systems to be not stuck on 5.10 and since it's disabled on other distributions, it seems that no one cares about this feature on 32-bit kernels.

**Comment by martinetd** [comment]
misuzu wrote on Fri, Jul 08, 2022 at 11:26:54PM -0700:
> I'm not a kernel specialist, but that's probably involves some kind of
> mailing list? I'm simply want my systems to be not stuck on 5.10 and
> since it's disabled on other distributions, it seems that no one cares
> about this feature on 32-bit kernels.

Yes, just send a mail to ***@***.*** and probably
***@***.*** in cc asking if it's a known bug and if someone is
working on it.

I'm not asking you to fix it, just reporting the problem with a little
bit of information (e.g. in what environment it happens) goes a long
way; the developpers probably just haven't used a 32bit machine in a
while...

I guess I can send the mail if you don't want to, but it's a bit odd if
it comes from someone who can't reproduce it -- from the looks of what's
said on discourse it doesn't happen with cross build and need a native
environment so I'd need to setup a 32bit VM or something first...

**Comment by vcunat** [comment]
Heh, GitHub redacted the addresses.

**Comment by martinetd** [comment]
geez, thanks for the reply I wouldn't have noticed. It looks like it works when I edit it and is only filtered if it comes from mail?
Anyway, if it didn't work: dwarves and bpf at vger.kernel.org

**Comment by misuzu** [comment]
> from the looks of what's said on discourse it doesn't happen with cross build and need a native environment so I'd need to setup a 32bit VM or something first...

Doing `nix-build . -A pkgsi686Linux.linux_5_15` on x86_64 machine is enough to trigger this issue.

**Comment by vcunat** [comment]
Hydra does it the same way, with x86_64 builders.  Then we have VM tests as well where a 32-bit kernel would be used, but this happens before during a regular build.

**Comment by martinetd** [comment]
I've triggered a build and I think I really need to mount something on /tmp... I've lost all remote access at some point during the build after growing the tmpfs and hoping it'd fit :/

Well, if it's reproducable as "cross" build (`ARCH=i386`) I'll get around to it eventually and can send the mail; thanks.

**Comment by martinetd** [comment]
*insert mandatory grmbling about linux build taking way too long because too many options are compiled in, and the usual difficulties to reproduce something manually after checking it fails in nix-build*

with that out of the way, it looks like a bug in our build of pahole 32bit builds.

```
/tmp/linux-5.15.52/build$ LLVM_OBJCOPY="objcopy" /nix/store/8fcywp2fd8zsx65b5kp94z0392drbgdy-pahole-1.23/bin/pahole/bin/pahole -J --btf_base vmlinux arch/x86/crypto/aesni-intel.ko
bash: /nix/store/8fcywp2fd8zsx65b5kp94z0392drbgdy-pahole-1.23/bin/pahole/bin/pahole: Not a directory

$ nix-store -qR /nix/store/8fcywp2fd8zsx65b5kp94z0392drbgdy-pahole-1.23/bin/pahole
/nix/store/7z2mlf9jqp7fpbb3wib75xdlgfi3bgy4-libunistring-1.0
/nix/store/nkvwq9wn2qv8gz2zxfdprdxdm4nfg93x-libidn2-2.3.2
/nix/store/pp1yzf2flqnjm02jnjcgxhx015bhmi9j-glibc-2.34-210
/nix/store/1y304j4nrcg4azbcfgsinjk5sbfyn807-libelf-0.8.13
/nix/store/4v7kprliykhgfm6r2y2amdr14qxpmlam-bzip2-1.0.6.0.2
/nix/store/6xx7safscsa8bmam6xawxngjh3fb9zk5-bash-5.1-p16
/nix/store/gxmvhzsi0nm17nlmpnvn7k3ycx6bbsgz-gcc-10.4.0-lib
/nix/store/n68dgb3rk2k51h22fmlf6jl3w61fax7w-zstd-1.5.2
/nix/store/p77ikss2xc161g0ahpixrq07nlg14ivs-zlib-1.2.12
/nix/store/wnxn3hjs1qa9nzvv06yc1j8ammjv3i0d-xz-5.2.5
/nix/store/gcm65k6iiw7yfbx53gpwqbajgcsv74bj-elfutils-0.187
/nix/store/nv146a9jbpqfxfch9g093pkcmhms1fl5-libbpf-0.7.0
/nix/store/8fcywp2fd8zsx65b5kp94z0392drbgdy-pahole-1.23

$ file /nix/store/8fcywp2fd8zsx65b5kp94z0392drbgdy-pahole-1.23/bin/pahole
/nix/store/8fcywp2fd8zsx65b5kp94z0392drbgdy-pahole-1.23/bin/pahole: ELF 32-bit LSB executable, Intel 80386, version 1 (SYSV), dynamically linked, interpreter /nix/store/pp1yzf2flqnjm02jnjcgxhx015bhmi9j-glibc-2.34-210/lib/ld-linux.so.2, for GNU/Linux 2.6.32, not stripped

```

with 64 bit version in same source tree it works:
```
/tmp/linux-5.15.52/build$ LLVM_OBJCOPY="objcopy" /nix/store/h6kzgddq80l5pkwgpi5x7hwwk73gyd1w-pahole-1.23//bin/pahole -J --btf_base vmlinux arch/x86/crypto/aesni-intel.ko
<no output, OK>

$ nix-store -qR /nix/store/h6kzgddq80l5pkwgpi5x7hwwk73gyd1w-pahole-1.23/
/nix/store/gi4zwjfcyqqwzzj6d76a1adfdzlrscjc-libunistring-1.0
/nix/store/fsxbvxgyycq9qh5bzkij573mj75hq7wp-libidn2-2.3.2
/nix/store/k56d9sk88pvrqhvwpa6msdf8gpwnimf6-glibc-2.34-210
/nix/store/34qw97f3h3rn5mrmq7l051g7ykyclg6x-libelf-0.8.13
/nix/store/3j18grljsyy4nxc078g00sy4cx6cf16g-bash-5.1-p16
/nix/store/hcdak2r7n3g850iw1hmhiasi0nzchdw8-zlib-1.2.12
/nix/store/6cq8w5jnib2pc9mcjb0ik59zyj1ysbz7-libbpf-0.7.0
/nix/store/bym6162f9mf4qqsr7k9d73526ar176x4-gcc-11.3.0-lib
/nix/store/p6vws9zzv997asjmrnyc33v9xiw5pjyr-xz-5.2.5
/nix/store/qh3ggnnzs34i3xmcvxbf3zr1504mmlbw-zstd-1.5.2
/nix/store/v014j9fmgn14jb7clkqy0a6n61nxapdw-bzip2-1.0.6.0.2
/nix/store/n4rvn2x09wkad3cjmikdk85g3ybd4pvq-elfutils-0.187
/nix/store/h6kzgddq80l5pkwgpi5x7hwwk73gyd1w-pahole-1.23

$ file /nix/store/h6kzgddq80l5pkwgpi5x7hwwk73gyd1w-pahole-1.23/bin/pahole
/nix/store/h6kzgddq80l5pkwgpi5x7hwwk73gyd1w-pahole-1.23/bin/pahole: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /nix/store/k56d9sk88pvrqhvwpa6msdf8gpwnimf6-glibc-2.34-210/lib/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, not stripped
```

the interesting part is that 32bit version of pahole works with 64bit binaries, it's only 32bit+32bit that seems broken?

Anyway, I've sent a mail to dwarves@, the thread is visible here: https://lore.kernel.org/dwarves/YsrXjggSxfe2U7Oz@codewreck.org/T/#u

**Comment by martinetd** [comment]
Geh. What a waste of time.
So this got fixed when I changed libbpf to build from libelf to elfutils back in #175467 last month -- which just got merged a couple of days ago (thanks @vcunat!). And I was testing on nixpkgs right before that merge...

Nothing to see, move along... I've opened #181031 to revert this but I don't really care about it ultimately as I don't use any 32bit system, happy to let this be forgotten. I mean, I'd love to see if there are more regressions on this, but if I'm not getting notified of these failures that's just making more work for everyone.

**Comment by anakryiko** [comment]
So the issue was in underlying libelf version, is that right? cc @acmel, seems like the issue was resolved

**Comment by martinetd** [comment]
Yes, I tested just changing from libelf to elfutils with libbpf 0.7.0 and that seemed to fix it for me as well; so I'd say it's a libelf bug.

my understanding is that libelf is a more generic (older, no longer maintained -- libelf 0.8.13 hasn't had any update since 2009!!!) implementation, and elfutils is somewhat linux-specific and came as a drop-in replacement (with elfutils-libelf) implementing extra features.
I had to change it for libbpf because it started using these extra features, but it looks like elfutils isn't as linux-specific as I thought so it might be worth trying to slowly deprecate libelf if we can get replacements running... At the very least I'm pretty sure using libelf for libbpf was an honest mistake, kernel builds usually require elfutils so I'd expect nobody in bpf-land expected libelf to be used.

**Comment by martinetd** [comment]
Ah, didn't realize who I was talking to, sorry: to clarify, I can confirm only changing the libelf provider from libelf to elfutils fixed it, keeping the libbpf/pahole versions identical.


@vcunat btw that repology page shows libelf isn't packaged on debian >= 9 (they're at 11 or 12? so dropped ~6 years ago) nor any rhel/fedora, stable suse or alpine > 3.9 release (wet-finger 3 years ago). That'd tend to say it's probably possible to get around without it :)

**Pull request merged**